tests/test_discrete_demography.py

```diff
@@ -0,0 +1,741 @@
+        G = 1.01
+        TEMP = EN0
+        # NOTE the range here. The off-by-1 thing
+        # can be tricky.
+        for i in range(0, 104-72):
+            EN0 = np.round(TEMP*np.power(G, i))
+        move = np.round(0.5*EN0)
+        EN0 -= move  # After movement from 0 to 2 at 103
+        EN2 += move
+        TEMP = EN0
+        # NOTE: the range again...
+        for i in range(0, 201-103):
+            EN0 = np.round(TEMP*np.power(G, i))
+        move = np.round(0.5*EN1)
+        EN2 += move
+        EN1 -= move
+
+        N0 = len(np.where(md['deme'] == 0)[0])
+        N1 = len(np.where(md['deme'] == 1)[0])
+        N2 = len(np.where(md['deme'] == 2)[0])
+        self.assertEqual(self.pop.N, N0+N1+N2)
+        self.assertEqual([N0, N1, N2], [EN0, EN1, EN2])
+
+    def testSimpleCopy(self):
+        Ninit = self.pop.N
+        states = [fwdpy11.PopulationState(self.pop.N),
+                  fwdpy11.PopulationState(0)]
+
+        configs = [fwdpy11.CopyIndividuals(50, 0, 1, 0.5)]
+        ss = fwdpy11.SoftSelection(states, configs)
+        ddr.SoftSelection_roundtrip(self.rng, self.pop, ss, 200)
+        md = np.array(self.pop.diploid_metadata, copy=False)
+        N0 = len(np.where(md['deme'] == 0)[0])
+        N1 = len(np.where(md['deme'] == 1)[0])
+        self.assertEqual(N0, Ninit)
+        self.assertEqual(N1, Ninit//2)
+
+    def testBottleneckWithMoveToSplitScenario(self):
+        """
+        At the time of the split, the parental
+        deme is copied exactly and then its
+        size is changed immediately.
+        """
+        states = [fwdpy11.PopulationState(self.pop.N),
+                  fwdpy11.PopulationState(0)]
+
+        configs = [fwdpy11.MoveIndividuals(50, 0, 1, 0.5)]
+        states[1].register_state_change(50, 10)
+        ss = fwdpy11.SoftSelection(states, configs)
+        ddr.SoftSelection_roundtrip(self.rng, self.pop, ss, 200)
+        md = np.array(self.pop.diploid_metadata, copy=False)
+        N0 = len(np.where(md['deme'] == 0)[0])
+        N1 = len(np.where(md['deme'] == 1)[0])
+        self.assertEqual(N0, 50)
+        self.assertEqual(N1, 10)
+
+    def testStandardBottleneckWithSplitScenario(self):
+        """
+        At the time of the split, the parental
+        deme is copied exactly and then its
+        size is changed immediately.
+        """
+        Ninit = self.pop.N
+        states = [fwdpy11.PopulationState(self.pop.N),
+                  fwdpy11.PopulationState(0)]
+
+        configs = [fwdpy11.CopyIndividuals(50, 0, 1, 1.0)]
+        states[1].register_state_change(50, Ninit//10)
+        ss = fwdpy11.SoftSelection(states, configs)
+        ddr.SoftSelection_roundtrip(self.rng, self.pop, ss, 200)
+        md = np.array(self.pop.diploid_metadata, copy=False)
+        N0 = len(np.where(md['deme'] == 0)[0])
+        N1 = len(np.where(md['deme'] == 1)[0])
+        self.assertEqual(N0, Ninit)
+        self.assertEqual(N1, Ninit//10)
+
+    def testMigration(self):
```

Comment:
This test fails because it starts a demographic model with a set of "states" that are simply not compatible with the population, thus bringing up a UI issue.  What to do about it?

Possibilities:

1. Require that simulations check for this and adjust metadata accordingly at start of sim.
2. Allow for populations to be constructed with list of population sizes.  This would require coordination:  the list would have to be "max demes" long.

Reply:
Proposition one seems the most straightforward, and we can fairly easily implement a simple API to handle the required ops.

Reply:
There is another option:

3. Stop requiring the user to construct the population and instead set it up properly (using something like option 2 above), evolve it, and then return it.

Reply:
Currently, we are going for option 1, which is perhaps the least interesting/useful of the bunch, but it'll let us make progress for now.